cms/app_registration.py

```diff
@@ -0,0 +1,142 @@
+    """
+    classes = []
+    # Find all classes that inherit from klass
+    for name, obj in inspect.getmembers(cms_module):
+        is_cms_config = (
+            inspect.isclass(obj) and
+            issubclass(obj, klass) and
+            # Ignore the import of klass itself
+            obj != klass
+        )
+        if is_cms_config:
+            classes.append(obj)
+    return classes
+
+
+def _find_config(cms_module):
+    """
+    Helper function.
+
+    Returns the class inheriting from CMSAppConfig in the given module.
+    If no such class exists in the module, returns None.
+    If multiple classes inherit from CMSAppConfig, raises
+    ImproperlyConfigured exception.
+    """
+    cms_config_classes = _find_subclasses(cms_module, CMSAppConfig)
+    if len(cms_config_classes) == 1:
+        return cms_config_classes[0]
+    elif len(cms_config_classes) > 1:
+        raise ImproperlyConfigured(
+            "cms_config.py files can't define more than one "
+            "class which inherits from CMSAppConfig")
+
+
+def _find_extension(cms_module):
+    """
+    Helper function.
+
+    Returns the class inheriting from CMSAppExtension in the given module.
+    If no such class exists in the module, returns None.
+    If multiple classes inherit from CMSAppExtension, raises
+    ImproperlyConfigured exception.
+    """
+    cms_extension_classes = _find_subclasses(cms_module, CMSAppExtension)
+    if len(cms_extension_classes) == 1:
+        return cms_extension_classes[0]
+    elif len(cms_extension_classes) > 1:
+        raise ImproperlyConfigured(
+            "cms_config.py files can't define more than one "
+            "class which inherits from CMSAppExtension")
+
+
+def autodiscover_cms_configs():
+    """
+    Find and import all cms_config.py files. Add a cms_app attribute
+    to django's app config with an instance of the cms config.
+    """
+    for app_config in apps.get_app_configs():
+        try:
+            cms_module = import_module(
+                '%s.%s' % (app_config.name, CMS_CONFIG_NAME))
+        except:
+            # If something in cms_config.py raises an exception let that
+            # exception bubble up. Only catch the exception if
+            # cms_config.py doesn't exist
+            if module_has_submodule(app_config.module, CMS_CONFIG_NAME):
+                raise
+        else:
+            config = _find_config(cms_module)
+            extension = _find_extension(cms_module)
+            # We are adding these attributes here rather than in
+            # django's app config definition because there are
+            # all kinds of limitations as to what can be imported
+            # in django's apps.py and leaving it to devs to define this
+            # there could cause issues
+            if config:
+                app_config.cms_config = config(app_config)
+            if extension:
+                app_config.cms_extension = extension()
```

Comment:
should we also pass app_config here for consistency?

Reply:
IMO no because nobody should really be accessing `app_config.cms_extension` anyway, so there's no need for `app_config.cms_extension.django_app`. They are two different classes so don't think they need the init to be the same.

Reply:
Ok